ols/utils/auth_dependency.py

```diff
@@ -0,0 +1,210 @@
+"""Manage authentication flow for FastAPI endpoints with K8S/OCP."""
+
+import logging
+
+import kubernetes.client
+from fastapi import HTTPException, Request
+from kubernetes.client.rest import ApiException
+
+from ols.utils import config
+
+logger = logging.getLogger(__name__)
+
+
+class K8sClientSingleton:
+    """Return the Kubernetes client instances.
+
+    Ensures we initialize the k8s client only once per application life cycle.
+    manage the initialization and config loading.
+    """
+
+    _instance = None
+    _api_client = None
+    _auth_api = None
+    _authorization_api = None
+
+    def __new__(cls):
+        """Create a new instance of the singleton, or returns the existing instance.
+
+        This method initializes the Kubernetes API clients the first time it is called.
+        and ensures that subsequent calls return the same instance.
+        """
+        if cls._instance is None:
+            cls._instance = super(K8sClientSingleton, cls).__new__(cls)
+            configuration = kubernetes.client.Configuration()
+
+            try:
+                try:
+                    kubernetes.config.load_kube_config(
+                        client_configuration=configuration
+                    )
+                except Exception:
+                    try:
+                        logger.warn(
+                            "kubeconfig file not found, trying in-cluster configuration"
+                        )
+                        kubernetes.config.load_incluster_config(
+                            client_configuration=configuration
+                        )
+                    except Exception:
+                        logger.warn(
+                            "Failed to load in-cluster Kubernetes configuration"
```

Comment:
we should report the exception text

Reply:
added the exception message to error logs